R/utils_render_rtf_redo.R

```diff
@@ -240,6 +240,134 @@ rtf_table <- function(rows) {
   rtf_raw(paste(unlist(rows), collapse = ""))
 }
 
+# This is the assumed width of the area within page
+# margins in a default Word document
+standard_width_twips <- 9468L
+
+twip_factors <-
+  c(
+    `in` = 1440, `pt` = 20, `px` = 15,
+    `cm` = 566.9291, `mm` = 56.69291, `tw` = 1
+  )
+
+parse_length_str <- function(lengths_vec) {
+
+  if (length(lengths_vec) == 1 && lengths_vec == "auto") {
+    lengths_vec <- paste0(standard_width_twips, "tw")
+  }
+
+  match <- regexec("^([0-9.-]+)(%|[a-z]+)$", lengths_vec)
+  match <- regmatches(lengths_vec, match)
+
+  vals <- vapply(match, function(m) {
+    if (length(m)) {
+      as.numeric(m[[2]])
+    } else {
+      NA_real_
+    }
+  }, numeric(1))
+
+  units <- vapply(match, function(m) {
+    if (length(m)) {
+      m[[3]]
+    } else {
+      NA_character_
+    }
+  }, character(1))
+
+  bad_values <- nzchar(lengths_vec) & is.na(vals)
+
+  if (any(bad_values)) {
+    stop(
+      "Some of the values supplied as absolute lengths cannot be interpreted:\n",
+      "* Use either of: `px`, `pt`, `in`, `cm`, `mm`, or `tw`.",
+      call. = FALSE
+    )
+  }
+
+  dplyr::tibble(value = vals, unit = units)
+}
+
+abs_len_to_twips <- function(lengths_df) {
+
+  lengths_df %>%
+    dplyr::left_join(
+      tibble::enframe(twip_factors, name = "unit", value = "conv"),
+      by = c("unit" = "unit")
+    ) %>%
+    dplyr::mutate(
+      value = ifelse(!is.na(conv), value * conv, value),
+      unit = ifelse(!is.na(conv), "tw", unit)
+    ) %>%
+    dplyr::select(value, unit)
+}
+
+# The `col_width_resolver_rtf()` function returns a vector of
+# RTF column widths in units of twips (numeric)
+#
+# `table_width` and `col_widths` should be character vectors with
+# a numeric portion followed by the unit; the first should be a
+# length-1 vector and the latter should be a vector with as many
+# elements as there are visible columns in the gt table
+col_width_resolver_rtf <- function(table_width,
+                                   col_widths,
+                                   n_cols) {
+
+  if (table_width == "auto" && is.null(col_widths)) {
+    return(rep(standard_width_twips / n_cols, n_cols))
+  }
+
+  table_width <- abs_len_to_twips(parse_length_str(table_width))
+
+  # Should stop if row count of `table_width` isn't 1
+  stopifnot(nrow(table_width) == 1)
```

Comment:
This should happen first, and I'd just check `length(table_width)` before it's converted to a table.

Reply:
Done (by you, actually).